executor/new_executor.go

```
@@ -0,0 +1,185 @@
+	otherFilter  ast.ExprNode
+	outter       bool
+	leftSmall    bool
+}
+
+func joinTwoRow(a *Row, b *Row) *Row {
+	return &Row{RowKeys: append(a.RowKeys, b.RowKeys...), Data: append(a.Data, b.Data...)}
+}
+
+func (e *HashJoinExec) getHashKey(exprs []ast.ExprNode) (string, error) {
+	vals := make([]types.Datum, 0, len(exprs))
+	for _, expr := range exprs {
+		v, err := evaluator.Eval(e.ctx, expr)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		vals = append(vals, v)
+	}
+	if len(vals) == 0 {
+		return "SingleHashGroup", nil
+	}
+	result, err := codec.EncodeValue([]byte{}, vals...)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return string(result), err
+}
+
+// Fields implements Executor Fields interface
+func (e *HashJoinExec) Fields() []*ast.ResultField {
+	return e.fields
+}
+
+// Close implements Executor Close interface
+func (e *HashJoinExec) Close() error {
+	e.hashTable = nil
+	return nil
+}
+
+// Next implements Executor Next interface
+func (e *HashJoinExec) Next() (*Row, error) {
+	if !e.prepared {
+		e.hashTable = make(map[string][]*Row)
+		for {
+			row, err := e.smallExec.Next()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if row == nil {
+				e.smallExec.Close()
+				e.prepared = true
+				break
+			}
+			matched := true
+			if e.smallFilter != nil {
+				matched, err = evaluator.EvalBool(e.ctx, e.smallFilter)
+			}
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if !matched {
+				continue
+			}
+			hashcode, err := e.getHashKey(e.smallHashKey)
+			if err != nil {
+				return nil, err
+			}
+			if rows, ok := e.hashTable[hashcode]; !ok {
+				e.hashTable[hashcode] = []*Row{row}
+			} else {
+				e.hashTable[hashcode] = append(rows, row)
+			}
+		}
+		e.prepared = true
+	}
+	for {
+		bigRow, err := e.bigExec.Next()
+		if err != nil {
+			return nil, err
+		}
+		if bigRow == nil {
+			e.bigExec.Close()
+			return nil, nil
+		}
+		hashcode, err := e.getHashKey(e.bigHashKey)
```

This error is not checked.